src/prepare.js

```diff
@@ -116,20 +107,23 @@ function prepareElement(element, errorHandler, context) {
   }
 }
 
-function prepare(element, options = {}, context = {}) {
+async function prepare(element, options = {}, context = {}) {
   const {
     errorHandler = (error) => {
       throw error;
     },
   } = options;
-  return prepareElement(element, errorHandler, context).then(
-    ([children, childContext, p]) =>
-      Promise.all(
-        React.Children.toArray(children)
-          .map((child) => prepare(child, options, childContext))
-          .concat(p)
-          .filter(Boolean),
-      ),
+  const [children, childContext, preparePromise] = await prepareElement(
+    element,
+    errorHandler,
+    context,
+  );
+
+  // Recursively run prepare on children, and return a promise that resolves once all children are prepared.
+  return Promise.all(
+    React.Children.toArray(children)
+      .map((child) => prepare(child, options, childContext))
+      .concat(preparePromise || []),
   );
```

Comment:
How's this @LudvigHz
I don't want to go in depth in a comment. It is a bad idea as comments are rarely maintained. This should point you in the right direction, and further understanding can be gained by reading the code.